ambry-rest/src/main/java/com/github/ambry/rest/HostQuotaEnforcementHandler.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.rest;
+
+import com.github.ambry.config.QuotaConfig;
+import com.github.ambry.quota.EnforcementRecommendation;
+import com.github.ambry.quota.QuotaManager;
+import com.github.ambry.quota.QuotaMode;
+import io.netty.buffer.Unpooled;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.codec.http.DefaultFullHttpResponse;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.netty.handler.codec.http.HttpUtil;
+import io.netty.handler.codec.http.HttpVersion;
+import java.util.ArrayList;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that enforces host level quota before processing requests.
+ */
+@ChannelHandler.Sharable
+public class HostQuotaEnforcementHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger logger = LoggerFactory.getLogger(HostQuotaEnforcementHandler.class);
+  private final static byte[] TOO_MANY_REQUESTS = "TOO_MANY_REQUESTS".getBytes();
+  private final NettyMetrics metrics;
+  private final QuotaManager quotaManager;
+  private final QuotaConfig quotaConfig;
+  private FullHttpResponse response;
+
+  /**
+   * Constructor for {@link HostQuotaEnforcementHandler}.
+   * @param nettyMetrics {@link NettyMetrics} object.
+   * @param quotaManager {@link QuotaManager} object.
+   * @param quotaConfig {@link QuotaConfig} object.
+   */
+  public HostQuotaEnforcementHandler(NettyMetrics nettyMetrics, QuotaManager quotaManager, QuotaConfig quotaConfig) {
+    this.metrics = nettyMetrics;
+    this.quotaManager = quotaManager;
+    this.quotaConfig = quotaConfig;
+  }
+
+  @Override
+  public void channelActive(ChannelHandlerContext ctx) throws Exception {
```

Comment:
I don't think this is the correct handler implementation. Maybe we can remove the handler implementation out from this PR and just get the various enforcer and quota manager ready.

1. channelActive is the callback when a tcp connection is established from client to frontend. we are not using http2 at frontend, so the channel is always a tcp channel.
2. HostLevelThrottler should be invoked when there is an http request, not when the connection is established.